pkgs/development/octave-modules/symbolic/default.nix

```diff
@@ -0,0 +1,39 @@
+{ buildOctaveLibrary
+, stdenv
+, fetchurl
+, octave
+, gnuplot
+, pythonEnv
+, texinfo
+}:
+
+(buildOctaveLibrary rec {
+  pname = "symbolic";
+  version = "2.9.0";
+
+  src = fetchurl {
+    url = "https://octave.sourceforge.io/download.php?package=${pname}-${version}.tar.gz";
+    sha256 = "1jr3kg9q6r4r4h3hiwq9fli6wsns73rqfzkrg25plha9195c97h8";
+  };
+
+  buildInputs = [ pythonEnv ];
+
+  meta = with stdenv.lib; {
+    homepage = "https://octave.sourceforge.io/${pname}/index.html";
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers; [ KarlJoad ];
+    description = "Adds symbolic calculation features to GNU Octave";
+  };
+}).overrideAttrs (oldAttrs: rec {
```

Comment:
Why building and then overriding? :)

Reply:
I was under the impression that `overrideAttrs` evaluates the original, then evaluates the override without a build in between. I came to that conclusion because of the manual section about [`overrideAttrs`](https://nixos.org/manual/nixpkgs/stable/#sec-pkg-overrideAttrs).

The reason I need to do this is because symbolic needs a way to reference the python environment that is built for it, and the standard `buildOctaveLibrary` doesn't seem to have support for overriding its various phases to achieve what I wanted.

Reply:
I needed the override because `symbolic` depends on being able to call `python3` from the environment (defined in `private/defaultpython.m`). Right now, I am using `substituteInPlace` to replace that reference. To do that:
1. I need to unpack the tarball (`dontUnpack = false`)
2. Perform the substitution (`buildPhase`)
3. Rebuild the tarball (`installPhase`)

Reply:
Something that I thought of while I was working on the octave's arduino package was that `propagatedBuildInputs` doesn't actually get used in any meaningful way. See [here](https://github.com/NixOS/nixpkgs/pull/108562#issuecomment-756554745) for a further discussion of the solution I would *like* to use.

If I can get `wrap.sh` to work correctly and add `propagatedBuildInputs` to the `PATH`, then I can completely remove the override. Because, if I have `propagatedBuildInputs = [ pythonEnv ];`, then the python interpreter for that environment will be added to `PATH`, and that interpreter has all the correct wrapping needed to find its declared libraries. This removes the need for me to mess around with the tarball at all.

Reply:
We should avoid using the "let's wrap everything that is in `propagatedBuildInputs` " hack in more places. If something needs to be wrapped, it best be explicit.

In https://github.com/NixOS/nixpkgs/pull/102613 I introduce `requiredPythonModules` to mark Python dependencies. Maybe you want the same here.

Reply:
Ok. I will work on switching what I have already to `requiredOctaveModules`, mimicking Python's runtime dependencies. I wanted to start by getting the "let's wrap everything that is in `propagatedBuildInputs` working first. That way, I at least have a sanity check.

Reply:
@FRidh I have introduced the `requiredOctavePackages` attribute and associated functions to ensure that octave packages have all their other octave package dependencies tracked separately from others.

I am deliberately leaving the "`propagatedBuildInputs` hack" in the code, because it is needed so that packages (like the python environment for `symbolic`) are properly referenced in wrappers for octave. I wanted to handle these separately because both of these are run-time dependencies, but octave package dependencies should actually be installed into Octave, like a package the user specifies. However, other runtime dependencies should not (and cannot, particularly now) be installed into octave.